labgrid/driver/shelldriver.py

```diff
@@ -139,9 +168,9 @@ def _await_login(self):
             )
 
             if index == 0:
-                if did_login and not did_silence_kernel:
+                if not did_silence_kernel:
```

Comment:
The `did_login` was an attempt to ensure that this only ran once per activation, otherwise it runs every time, which isn't desirable. I _think_ this might actually be a bug in that `self.status` on line 178 below should be `self._status` so that the code in `on_activate` will only run once per activation.

Perhaps one of the Labgrid devs can clarify

Reply:
> The `did_login` was an attempt to ensure that this only ran once per activation

Shouldn't that `did_silence_kernel` be enough for that purpose? I never hit that `did_login=True` code path as there is no `login_prompt` on my system, so this is actually a regression for me, introduced in 40154266ba3.

See it yourself, with the following changes applied:

```diff
diff --git a/labgrid/driver/shelldriver.py b/labgrid/driver/shelldriver.py
index aeb3e54..7b54190 100644
--- a/labgrid/driver/shelldriver.py
+++ b/labgrid/driver/shelldriver.py
@@ -166,21 +166,25 @@ class ShellDriver(CommandMixin, Driver, CommandProtocol, FileTransferProtocol):
                 expectations,
                 timeout=self.await_login_timeout
             )
+            self.logger.info(f"shelldriver: while loop, index={index}")
 
             if index == 0:
                 if not did_silence_kernel:
                     # Silence the kernel and wait for another prompt
+                    self.logger.info("shelldriver: silencing kernel")
                     self._silence_kernel(timeout)
                     did_silence_kernel = True
                 else:
                     # we got a prompt. no need for any further action to
                     # activate this driver.
+                    self.logger.info("shelldriver: got a prompt, self.status=1")
                     self.status = 1
                     break
 
             elif index == 1:
                 # we need to login
                 self.console.sendline(self.username)
+                self.logger.info("shelldriver: did login, did_login=True")
                 did_login = True
 
             elif index == 2:
@@ -190,8 +194,10 @@ class ShellDriver(CommandMixin, Driver, CommandProtocol, FileTransferProtocol):
                     raise Exception("Password entry needed but no password set")
 
             elif index == 3:
+                self.logger.info("shelldriver: expect hit a timeout")
                 # expect hit a timeout while waiting for a match
                 if before == last_before:
+                    self.logger.info("shelldriver: expect did not received anything, sending empty line")
                     # we did not receive anything during
                     # self.await_login_timeout.
                     # let's assume the target is idle and we can safely issue a
@@ -199,6 +205,7 @@ class ShellDriver(CommandMixin, Driver, CommandProtocol, FileTransferProtocol):
                     self.console.sendline("")
 
             elif index == 4:
+                self.logger.info("shelldriver: console activated")
                 # we have just activated a console here
                 # lets start over again and see if login or prompt will appear
                 # now.
@@ -213,6 +220,7 @@ class ShellDriver(CommandMixin, Driver, CommandProtocol, FileTransferProtocol):
             self.console.settle(self.post_login_settle_time, timeout=timeout.remaining)
 
         if did_login:
+            self.logger.info("shelldriver: did_login=True, executing _check_prompt()")
             self._check_prompt()
 
     @step()
```

I'm seeing following:

```shell
   INFO: shelldriver: while loop, index=4
   INFO: shelldriver: console activated
   INFO: shelldriver: while loop, index=0
   INFO: shelldriver: silencing kernel
   INFO: shelldriver: while loop, index=3
   INFO: shelldriver: expect hit a timeout
   INFO: shelldriver: while loop, index=3
   INFO: shelldriver: expect hit a timeout
   INFO: shelldriver: expect did not received anything, sending empty line
   INFO: shelldriver: while loop, index=0
   INFO: shelldriver: got a prompt, self.status=1
```

Reply:
> I think this might actually be a bug in that self.status on line 178 below should be self._status so that the code in on_activate will only run once per activation.

Indeed, this looks like a bug to me as well. Anyway, in your use case, when you've console with login, you should get [`self._status = 1`](https://github.com/labgrid-project/labgrid/blob/5d78a63b27b281b99f23971e597966d0c29a1751/labgrid/driver/shelldriver.py#L207) set via `_check_prompt()`:

https://github.com/labgrid-project/labgrid/blob/5d78a63b27b281b99f23971e597966d0c29a1751/labgrid/driver/shelldriver.py#L183-L186

Reply:
Ah, I missed that `_check_prompt` did that also. Maybe the line in `self.status` line in `_await_login` should just be removed?

Reply:
That `self.status` line doesn't do anything since the variable that is used by the driver is `self._status`.

Reply:
So I've fixed it by renaming it.